stackit/internal/services/resourcemanager/project/resource.go

```diff
@@ -478,3 +660,104 @@ func toUpdatePayload(model *Model) (*resourcemanager.PartialUpdateProjectPayload
 		Labels:            labels,
 	}, nil
 }
+
+// updateMembers adds and removes members to match the model
+func updateMembers(ctx context.Context, projectId string, modelMembers *[]authorization.Member, client *authorization.APIClient) error {
+	if modelMembers == nil || len(*modelMembers) == 0 {
+		return nil
+	}
+
+	// Get current members
+	currentMembersResp, err := client.ListMembersExecute(ctx, projectResourceType, projectId)
+	if err != nil {
+		return fmt.Errorf("get members: %w", err)
+	}
+
+	type memberState struct {
+		isInModel bool
+		isCreated bool
+		subject   string
+		role      string
+	}
+
+	membersState := make(map[string]*memberState) // Key in the form of "subject,role"
+	for _, m := range *modelMembers {
+		mId := memberId(m)
+		membersState[mId] = &memberState{
+			isInModel: true,
+			subject:   *m.Subject,
+			role:      *m.Role,
+		}
+	}
+
+	for _, m := range *currentMembersResp.Members {
+		if utils.IsLegacyProjectRole(*m.Role) {
+			continue
+		}
+
+		mId := memberId(m)
+		_, ok := membersState[mId]
+		if !ok {
+			membersState[mId] = &memberState{}
+		}
+		membersState[mId].isCreated = true
+		membersState[mId].subject = *m.Subject
+		membersState[mId].role = *m.Role
+	}
+
+	// Add/remove members
+	membersToAdd := make([]authorization.Member, 0)
+	membersToRemove := make([]authorization.Member, 0)
+	for _, state := range membersState {
+		if state.isInModel && !state.isCreated {
+			m := authorization.Member{
+				Subject: &state.subject,
+				Role:    &state.role,
+			}
+			membersToAdd = append(membersToAdd, m)
+
+			infoMsg := fmt.Sprintf("### Will add member to project: { role: %s, subject: %s }", state.role, state.subject)
+			tflog.Warn(ctx, infoMsg)
+		}
+
+		if !state.isInModel && state.isCreated {
+			m := authorization.Member{
+				Subject: &state.subject,
+				Role:    &state.role,
+			}
+			membersToRemove = append(membersToRemove, m)
+
+			infoMsg := fmt.Sprintf("### Will remove member from project: { role: %s, subject: %s }", state.role, state.subject)
+			tflog.Warn(ctx, infoMsg)
+		}
+	}
+
+	if len(membersToAdd) > 0 {
+		payload := authorization.AddMembersPayload{
+			Members:      &membersToAdd,
+			ResourceType: sdkUtils.Ptr(projectResourceType),
+		}
+		_, err := client.AddMembers(ctx, projectId).AddMembersPayload(payload).Execute()
+		if err != nil {
+			return fmt.Errorf("add members: %w", err)
+		}
+	}
+
+	if len(membersToRemove) > 0 {
+		payload := authorization.RemoveMembersPayload{
+			Members:      &membersToRemove,
+			ResourceType: sdkUtils.Ptr(projectResourceType),
+		}
+		_, err := client.RemoveMembers(ctx, projectId).RemoveMembersPayload(payload).Execute()
+		if err != nil {
+			return fmt.Errorf("remove members: %w", err)
+		}
+	}
```

Comment:
Good job on the logic!

As mentioned previously, I'm worried about the terraform state if any of these fail. Here we could set the state inside these methods and between calls. It should also address the previous comments

Reply:
I think to update the state after adding the members we would need to make an extra API call anyway (list members), which could also fail